app/services/analysis/gwas/genotype_hapmap_to_csv_converter.rb

```diff
@@ -0,0 +1,147 @@
+
+          raise unless hapmap_data.valid?
+
+          mutation_names = []
+          mutation_data = []
+          removed_mutation_names = []
+          sample_data = Hash.new { |data, sample_name| data[sample_name] = [] }
+
+          hapmap_data.each_record do |record|
+            record_status, record_data = process_record(record)
+
+            case record_status
+            when :ok
+              mutation_names << record_data[:mutation_name]
+              mutation_data << record_data[:mutation_data]
+
+              record_data[:sample_values].each do |sample_name, value|
+                sample_data[sample_name] << value
+              end
+
+            when :removed
+              removed_mutation_names << record_data[:mutation_name]
+            end
+          end
+
+          [
+            :ok,
+            generate_genotype_csv(mutation_names, sample_data),
+            generate_map_csv(mutation_names, mutation_data),
+            removed_mutation_names
+          ]
+        end
+      end
+
+      private
+
+      # Extract mutation and samples data from a single hapmap record. One record
+      # holds information about one mutation and multiple samples.
+      #
+      # If there is no variation (i.e. there is only one distinct value apart
+      # from NA for each sample) for the mutation then record is silently skipped.
+      #
+      # Special characters apart from underscore are stripped from sample and
+      # mutation identifiers.
+      #
+      # record - hapmap record
+      #
+      # Returns a hash containing mutation name, position and value for each sample.
+      def process_record(record)
+        mutation_name = record.rs.strip.gsub(/\W/, '_')
+        mutation_data = [record.chrom, record.pos]
+        sample_values = []
+
+        record.sample_ids.each.with_index do |sample_name, sample_idx|
+          value = sample_value(record.alleles, record.samples[sample_idx])
+
+          sample_values << [sample_name, value]
+        end
+
+        unique_values = sample_values.map { |_, val| val }.uniq
+
+        return [:removed, { mutation_name: mutation_name }] if (unique_values - ["NA"]).size <= 1
+
+        [:ok, {
+          mutation_name: mutation_name,
+          mutation_data: mutation_data,
+          sample_values: sample_values
+        }]
+      end
+
+      # Find number of alternative alleles for given sample.
+      #
+      # mutation - definition of mutation (e.g. A/C)
+      # sample - sample alleles (e.g. AA, AC, CC, NN)
+      #
+      # Returns:
+      #   2 for alternative allele in each chromosome
+      #   1 for alternative allele in any chromosome
+      #   0 for no mutation
+      #   NA if call could not be made
+      def sample_value(mutation, sample)
+        verify_sample(sample)
+
+        _original, alternative = mutation.split("/")
+
+        return "NA" if alternative == "NN"
+
+        alternative_count = sample.split("").count { |allele| allele == alternative }
```

Comment:
I must say I am a bit surprised you need to use `split` in such a strange way to turn a string into an enumerable. But it looks like there really is no other (simple) way... :)

Reply:
It appears there is a better way: `sample.chars`.